### PR TITLE
[JENKINS-29254] Use fabric8 as Kubernetes client library

### PR DESCRIPTION
**Pull request by ndeloof**
@reviewbybees

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 1]
```diff
@@ -323,7 +318,7 @@ public Node call() throws Exception {
                 }
 
                 return slave;
-            } catch (Exception ex) {
+            } catch (Throwable ex) {
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: isn't catching Error a bad idea?

File: pom.xml [hunk 2]
```diff
@@ -50,11 +51,84 @@
...
     <dependency>
-      <groupId>org.csanchez.kubernetes</groupId>
-      <artifactId>kubernetes-java</artifactId>
-      <version>0.0.3</version>
+      <groupId>io.fabric8</groupId>
+      <artifactId>kubernetes-api</artifactId>
+      <version>2.2.5</version>
+    </dependency>
+
+    <!-- Required to prevent a version conflict with jenkins-core NoSuchMethodError:org.springframework.util.ClassUtils.isCglibProxyClass -->
```
Comment (carlossg) [inline pom.xml]: if they are already included transitively should they be in dependencyManagement ?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 3]
```diff
@@ -425,6 +429,6 @@ public FormValidation doTestConnection(@QueryParameter URL serverUrl, @QueryPara
 
     @Override
     public String toString() {
-        return MoreObjects.toStringHelper(this).add("name", name).add("serverUrl", serverUrl).toString();
+        return "KubernetesCloud '" + name + "' serverUrl :" + serverUrl;
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: more of a personal taste question, but why don't you like Guava MoreObjects ?

**Comment by carlossg** [comment]
If you tried it against Google Container Engine and works I'm ok. Just added 3 comments

File: pom.xml (hunk 2 above)
Comment (ndeloof) [inline pom.xml]: I've got strange behavior with dependencies not being packaged so preferred to make it explicit

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 1 above)
Comment (ndeloof) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: I got strange low level issues while experimenting so preferred to catch all throwable so they get logged

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 3 above)
Comment (ndeloof) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: Just didn't know where this class came from as I changed dependencies, so opted for simplest solution. I'm also not a big fan for those util library to create toString, such a simple thing to generate from IDE without being more verbose

Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: let's agree to disagree, I prefer to use helpers and consistent helper methods without code generation

**Comment by carlossg** [comment]
:bee:

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 3 above)
Comment (ndeloof) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: My point is those are useful when they significantly make code less verbose, but can't see significant benefit here. Also always wonder about the actual performance impact for 5 method calls (even I know String concat is not a good sample about taking care for performances)

**Comment by carlossg** [comment]
if that works LGTM

**Comment by KostyaSha** [comment]
What is the rule of mixing shaded and not shaded object imports?

**Comment by ndeloof** [comment]
@KostyaSha not intentional

**Comment by KostyaSha** [comment]
@ndeloof be careful because it may cause issues.

**Pull request merged**